ebpf unit testing

xinyuanzzz opened the pull request:
Considering that eBPF programs are in essence C programs, we propose to leverage existing C-native testing infrastructure to dramatically improve eBPF testing experience and enhance productivity. To this end we generate mock interfaces for kernel-based helper functions and also customized functions in Cilium with CMock. By including the generated mock module, eBPF programs can be easily and conveniently unit-tested as normal C programs.

We demonstrate the approach by performing a sample unit test on functions in
lib/nat.h. The test functions are written in nat_test.h.

To make the demo, run "make -C test/bpf nat-test" to produce an executable file.

As we briefly introduced this work in the Cilium dev sync session on July 7, we would like to get feedbacks, especially on which funtions in Cilium need this coverage.

Signed-off-by: Xinyuan Zhang <zhangxinyuan@google.com>

joestringer commented on bpf/mock/helpers_skb.h:
```diff
@@ -0,0 +1,106 @@
+// If other helper functions are needed, please add the function declarations at the bottom.
+// After adding, generate mock helper functions again with "$ruby cmock.rb -obpf.yaml helpers_skb.h".
+
+#ifndef __MOCK_HELPERS_SKB__
```
Some of these headers seem like they're duplicated or derived from the existing versions. Is there a way to avoid duplication? How were these created? How do we keep them up to date?

xinyuanzzz replied:
You are right, bpf/mock/helpers_skb.h and bpf/mock/helpers_xdp.h are derived from bpf/helpers.h, bpf/helpers_skb.h, and bpf/helpers_xdp.h. It could be possible to avoid these headers, but we will have to modify a lot in the Cilium source files. It will be much eaiser to list the declarations of the helper functions in these headers because they are only used to be included and compiled with mock library. And if there are any additional helper functions needed in the future, just add the declarations at the bottom.

joestringer replied:
OK, but we should make it super clear how to update these (for instance, autogenerating this file via a make target, or having some checker that runs in CI that makes sure that this file is in sync with the one it's derived from and complains if they're not synchronized). Right now roughly ~none of the core team knows how this framework works, and I'd like to minimize the set of struggles to keep it updated & allow folks to use it in future. Otherwise it rots in the tree and when someone tries to use it, they have to solve two problems - updating the mock framework and then writing their tests. The fewer barriers to participate in this framework, the better.

xinyuanzzz replied:
Noted. I will try to get a way to generate those headers automatically.

Weil0ng replied:
Are you going to generate the full set automatically? If so, as Joe suggested, I think it makes sense to also add a target in the Makefile to generate the headers and a test in CI to check for parity between these headers and the existing kernel ones.

joestringer commented on bpf/mock/mock_conntrack_vacant.c:
```diff
@@ -0,0 +1,650 @@
+/* AUTOGENERATED FILE. DO NOT EDIT. */
```
If these files were autogenerated, do we need to commit them into the tree or can we just generate them at test time? I'd prefer the latter so we have fewer files to manage or keep up to date or ensure developers understand how to generate them correctly.

joestringer replied:
Oh and if we're worried about tooling not being available to generate & run the tests if the mock framework requires something unusual, the usual approach we use is just to wrap the required tooling into a docker container and mount the host filesystem into there to bootstrap.

xinyuanzzz replied:
Make sense! I suggest we keep the generated mock library for helper functions, assuming that the list of helpers functions does not change often. We can bypass the generated mock library for customized functions, and developers can generate them upon using.

joestringer commented on bpf/mock/conntrack_vacant.h:
```diff
@@ -0,0 +1,12 @@
+// Header of functions to be mocked in conntrack.h.
```
For each of the hand-written files, please add the license. The common form looks like this for BPF code:

```
/* SPDX-License-Identifier: GPL-2.0 */
/* Copyright (C) 2021 Authors of Cilium */
```

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -575,6 +575,7 @@ static __always_inline void ct4_cilium_dbg_tuple(struct __ctx_buff *ctx, __u8 ty
 	cilium_dbg(ctx, type, addr, rev_nat_index);
 }
 
+#ifndef __TEST__
```
Do we have to add these wrappers around every function that we test?

It'd be nice if we could separate out the declaration vs. implementation somehow and avoid littering the code with these macros for each tested function.

xinyuanzzz replied:
If we only mock helper functions, we can avoid that. But if developers want mock a specific customized function and not affect other functions in this file, they can wrap this function with such macros upon testing to avoid this function to be defined here (need to be defined in the mock library). We do not actually need to commit them into the tree, just a demo showing the necessary preparation before performing test.

joestringer replied:
Could the specific test just do `#define` trickery instead in the test? Basically forward-declare their mock version of the function then `#define ct_update mock_ct_update`? Then it can be up to the individual test on how it wants to handle this, instead of requiring the developer to understand which tests mock out which functionality and conditionally enable/disable some of the functions.

joestringer replied:
^^ This is basically what `bpf/tests/conntrack_test.h` does for the helper functions.

xinyuanzzz replied:
I tried this approach before but failed. Since CMock generates not only a mock header but also a mock c file to be compiled together, we have no way to replace a static inline function with a mock substitute using a macro trick like conntrack_test.h does. But I am still thinking of if there is a more elegant way to mock customized functions :).

joestringer replied:
Hmm okay, that's a bit awkward then. I don't have any other immediate suggestions, only a concern about this approach below.

> We do not actually need to commit them into the tree, just a demo showing the necessary preparation before performing test.

^ This statement seems to assume that testing is a manual process, but that should not be the case. If we require manual changes to the tree to be able to run a test, then no-one will run the test. These unit tests need to be automated such that we can run them regularly as part of CI testing pipelines. Also if a developer wants to run a test, they just run some make target locally which runs the test for them. 

Maybe this is not an immediate concern, but let's say you wanted to also unit-test `ct_lookup()`. The logic is not entirely trivial as it's not just a hash table lookup. There's IP fragment handling logic in there; there's different logic depending on the type of the packet and depending on the state in the conntrack table, etc. So I could see a desire to unit-test this logic. Separately, it looks like from this overall PR that you wanted to mock this functionality out so that you don't have to worry about bugs in ct_lookup() when testing nat.h. How can we let both tests coexist in the tree and regularly run them without manual input?

Weil0ng replied:
I think these unit tests could run at different granularity/abstraction level. For example, current nat tests might focus on logic in NAT itself without worrying about bugs in `ct_lookup()` and that's why we would like to mock it out. But I can imagine we could have another testsuite that covers the logic in `ct_lookup()` itself, such a test will be compiled without using current mock header. Essentially we could have different test files with different sets of real/mock header combinations determined by what they are testing. Does this make sense?

If so then we kinda need a way to substitute these functions (e.g., `ct_lookup()`) when needed with generated mocks, not sure if there's some compiler trick we can use to achieve that purpose?

joestringer replied:
:100: exactly my thoughts @Weil0ng . Does CMock provide some kind of functionality to help with this? It feels like this kind of "unit testing at different layers of abstraction" problem should be common across various testsuites using CMock. BPF style of defining everything in headers might make things slightly more complicated, but hopefully we can still come up with a solution.

xinyuanzzz replied:
This problem has been solved. We can actually replace the function we want to mock like conntrack_test.h does without modifying the source file. I wrongly intrepreted an error at the first attempt. Thanks for the suggestions!

joestringer reviewed (changes requested):
Cool stuff. I have just a few boring process / maintenance type bits of feedback. I'll leave the NAT test review up to someone with more familiarity in that area.

Weil0ng commented on bpf/mock/check_headers.sh:
```diff
@@ -0,0 +1,27 @@
+# script to check if helpers.h, helpers_skb.h, and helpers_xdp.h are up to date
+
+#!/bin/bash
+
```
Can you add some comment describing these operations?

Weil0ng commented on bpf/mock/check_headers.sh:
```diff
@@ -0,0 +1,27 @@
+
+if cmp helpers_temp.h helpers.h
+then echo mock/helpers.h is up to date!
+else echo mock/helpers.h is out of date! Please run generate_headers.sh or manually add the new helpers to mock/helpers.h and run mock_helpers.sh to generate mock libraries.
```
maybe we can add a Makefile target to call these scripts and then devs can simple run "make XXX"?

joestringer replied:
right, I think that the `sed` portions of this can potentially stay in a `.sed` file but the comparison / make pieces seem like they could be folded into a make target.

Weil0ng commented on bpf/mock/conntrack_vacant.h:
```diff
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
This file is now generated, correct? Can you inject a comment in these generated files? something like:
`/* GENERATED BY ... */`

xinyuanzzz replied:
This is not generated. For customized functions, developers need to create this header manually and choose which functions to be mocked.

Weil0ng commented on bpf/mock/demo.sh:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: GPL-2.0
```
Can you rename this file? I don't think we need to check in demo code, maybe repurpose this to be a general CMock generation script?

Weil0ng commented on bpf/mock/generate_headers.sh:
```diff
@@ -0,0 +1,10 @@
+# script to auto-generate helpers.h, helpers_skb.h, and helpers_xdp.h from bpf/helpers.h, bpf/helpers_skb.h, and bpf/helpers_xdp.h
+
+#!/bin/bash
+
```
ditto, would be great if you could add some comment describing the desired operation

Weil0ng commented on bpf/mock/helpers.h:
```diff
@@ -0,0 +1,84 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
This is now generated, correct?

xinyuanzzz replied:
Right!

joestringer replied:
If it's generated, do we still need to check this one in?

xinyuanzzz replied:
I would suggest we keep them because it could be possible that a future change on bpf/helpers.h may disable the script from generating the right headers (could be some lines wrongly trigger sed commends). If we keep them, at least developers can revise the headers manually.

joestringer replied:
I think we should commit to either autogenerating or manually generating, not mixing the two. And my preference would be on autogenerating. Otherwise as soon as one manual change is introduced, the automatic generation piece is broken and useless.

I see the concern though, for instance how will someone in future understand that the headerfile is not being generated correctly? Can we make that more obvious somehow? I'm not sure, as I don't have a good sense for what it looks like for a developer to work on a new test and hit an issue where this autogenerated header doesn't work.

xinyuanzzz replied:
Make sense, but I think there are another two reasons to keep them. First, the list of helper functions does not change often, so developers do not really need to generate the helper headers and generate the mock libraries for helpers, instead they can plug and play if we keep them. Second, even if the future add-ons cause some trouble, developers probably could use the old version mock libraries because they might not even need the newly added helpers.

joestringer replied:
If they aren't changing often and the autogenerated headers (or scripts to generate them) don't bitrot quickly then it seems fair enough. It'll save a bit of time either setting up a docker container to handle mock generation or forcing developers to install another new tool. I think we do keep some other generated code in the tree in similar situations to simplify the compilation process.

> Second, even if the future add-ons cause some trouble, developers probably could use the old version mock libraries because they might not even need the newly added helpers.

Only thing I want to point out here is that this starts by assuming that things are not maintained and then provides a workaround for that use case. I would argue that we should instead address this when those 'add-ons' are added to the tree (I assume you mean synchronizing the helpers header with upstream Linux) so that other developers don't have this problem in the first place. We can add tests for this into the CI so that any time the affected files are modified, we run specific mock generation + mock compile tests to validate that things aren't randomly breaking for developers later on when they try to write tests.

xinyuanzzz replied:
I agree with you.

> We can add tests for this into the CI so that any time the affected files are modified, we run specific mock generation + mock compile tests to validate that things aren't randomly breaking for developers later on when they try to write tests.

I think we can add a CI test to check if the headers need to update, if there are any change made to bpf/helpers, then it will raise a fail and ask the developer editing to generate the headers again (or revise the sed if needed).

Weil0ng reviewed (changes requested):
Would be great to add a README under bpf/mock/ to describe the steps to create/run unit tests :)

Weil0ng commented on bpf/mock/helpers.sed:
```diff
@@ -0,0 +1 @@
+1 s/^/\/\* GENERATED FROM lib\/helpers\.h \*\/\n/; /#ifndef BPF_/,/#endif/d; /#if __ctx_is/,/#endif/d; s/"compiler\.h"/<bpf\/compiler\.h>/g; /ctx\/ctx\.h/d; s/__BPF_HELPERS\(.*__\)/__MOCK_HELPERS\1/g; s/static //g; s/BPF_\w*(\(\w*\), /\1(/g; s/BPF_\w*(\(\w*\)/\1(/g; s/__printf(.*) //g; / =/{ N; s/ =.*;/;/ };
```
Can you add copyright info and comment that describes the operation?

xinyuanzzz replied:
Sorry, forgot it.

Weil0ng commented on bpf/mock/helpers_skb.sed:
```diff
@@ -0,0 +1 @@
+1 s/^/\/\* GENERATED FROM lib\/helpers_skb\.h \*\/\n/; /#ifndef BPF_/,/#endif/d; /#if __ctx_is/,/#endif/d; s/"features_skb\.h"/<bpf\/features_skb\.h>/g; s/"compiler\.h"/<bpf\/compiler\.h>/g; /"helpers\.h"/d; s/__BPF_HELPERS\(.*__\)/__MOCK_HELPERS\1/g; s/static //g; s/BPF_\w*(\(\w*\), /\1(/g; s/BPF_\w*(\(\w*\)/\1(/g; s/__printf(.*) //g; / =/{ N; s/ =.*;/;/ };
```
ditto

Weil0ng commented on bpf/mock/helpers_xdp.sed:
```diff
@@ -0,0 +1 @@
+1 s/^/\/\* GENERATED FROM lib\/helpers_xdp\.h \*\/\n/; /#ifndef BPF_/,/#endif/d; /#if __ctx_is/,/#endif/d; s/"features_xdp\.h"/<bpf\/features_xdp\.h>/g; s/"compiler\.h"/<bpf\/compiler\.h>/g; /"helpers\.h"/d; s/__BPF_HELPERS\(.*__\)/__MOCK_HELPERS\1/g; s/static //g; s/BPF_\w*(\(\w*\), /\1(/g; s/BPF_\w*(\(\w*\)/\1(/g; s/__printf(.*) //g; / =/{ N; s/ =.*;/;/ };
```
ditto

joestringer commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,44 @@
+cmock_dir = "../../../../cmock/lib/cmock.rb"
```
This path seems to go outside the repository at a very specific filepath which may or may not match other people's directory layouts. It also is called `cmock_dir` which suggests it's a directory, but it seems to be actually a binary path.

I would suggest declaring this like
```
CMOCK ?= cmock.rb
```

So this way, if cmock.rb is in your path, it will work. If it is not in your path, you can do `make -C bpf/mock CMOCK=../../../../cmock/lib/cmock.rb` and it should use that specific version of `cmock.rb`.

joestringer commented on bpf/mock/README.md:
````diff
@@ -0,0 +1,70 @@
+containing the declarations of the functions to be mocked, then run
+
+```bash
+make -C bpf/mock CMOCK=YOUR_cmock.rb_PATH mock_cutomized filename=NAME_OF_THE_HEADER_TO_BE_MOCKED
````
nit, there's a typo here, `mock_cutomized`. I see the same elsewhere in this file too.

joestringer commented on bpf/mock/README.md:
```diff
@@ -0,0 +1,70 @@
+## Dependencies
+
+[CMock](https://github.com/ThrowTheSwitch/CMock)
```
Probably makes sense to call out Ruby as a dependency here as well, not sure whether it comes standard in every environment.

joestringer commented on test/bpf/cmock/cmock.c:
```diff
@@ -0,0 +1,217 @@
+/* ==========================================
```
Which version did this come from? Can you point to a CMock release version(s) or commit sha(s) in the commit message?

joestringer reviewed (commented):
A few more minor comments below. I think we should still also think about how to integrate this into automated testing (current tests are run via test/bpf/unit-test). I'd like to also convert the existing conntrack tests over to this framework to get a good feel for how the framework works, but I don't have a lot of cycles for that in the short term. That shouldn't be a requirement to merge this PR, but would be a great way to gather feedback on this framework.

errordeveloper commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,59 @@
+# Copyright 2017-2020 Authors of Cilium
+# SPDX-License-Identifier: Apache-2.0
+
+CMOCK ?= cmock.rb
```
It'd be helpful to clarify what this file is, I don't see in this PR and from a quick search it appears to be part of CMake. At first I wasn't sure, and wondered if it was a new invention :)

errordeveloper commented on bpf/mock/README.md:
```diff
@@ -0,0 +1,71 @@
+## Dependencies
+
```
It'd help to add a few words here about what these dependencies are and why they are used, CMake and `cmock.rb` are new things to Cilium project itself. Some folks to use Ruby because of Vagrant, but broadly it's not a widely common dependency for Cilium developers.

errordeveloper reviewed (commented):
Personally, I hugely welcome this! We've discussed testing BPF code by loading it on some kind micro VMs, but in my knowledge the progress on that had been somewhat limited. It's probably still very beneficial to test on VMs, but having these unit will be an exciting leap forward!

errordeveloper commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,59 @@
+# Copyright 2017-2020 Authors of Cilium
```
Probably should use 2021 here and in other license headers :)

xinyuanzzz replied:
Thanks! I changed the license for bpf/mock/Makefile. For mock/helpers.h, mock/helpers_skb.h and mock/helpers_xdp.h, they are generated from the original files in include/bpf/, so their licenses are consistent.

Weil0ng reviewed (approved):
Framework LGTM!

Would be great to see a more thorough test for nat.h but I think we can probably do it in a follow up PR. This can focus on the framework and utilities.

errordeveloper commented on bpf/mock/helpers_xdp.sed:
```diff
@@ -0,0 +1,15 @@
+# 9&10 Restruct the functions to normal styles.
+# 11 Remove remappings.
+
+1 s/^/\/\* GENERATED FROM bpf\/helpers_xdp\.h \*\/\n/; /#ifndef BPF_/,/#endif/d; /#if __ctx_is/,/#endif/d; s/"features_xdp\.h"/<bpf\/features_xdp\.h>/g; s/"compiler\.h"/<bpf\/compiler\.h>/g; /"helpers\.h"/d; s/__BPF_HELPERS\(.*__\)/__MOCK_HELPERS\1/g; s/static //g; s/BPF_\w*(\(\w*\), /\1(/g; s/BPF_\w*(\(\w*\)/\1(/g; s/__printf(.*) //g; / =/{ N; s/ =.*;/;/ };
```
Seems like this could be spread over multiple lines? It'd make it much more readable if it wasn't all on one line. Since there are quite a few `/`s, I'd use a `|` and avoid extra escapes to make it also a bit easier to read, e.g. `s|"features_xdp\.h"|<bpf/features_xdp\.h>|`.

xinyuanzzz replied:
Fixed!

kkourt reviewed (commented):
Thanks, this is awesome!

I've checked out the branch to try and run the tests.

I'm running (based on the Run Unit Tests section):
> make -C test/bpf nat-test

But I get:
> make: Entering directory '/home/kkourt/src/cilium-trees/cilium-cmock/test/bpf'
clang -I../../bpf/ -I../../bpf/include -I. -D__NR_CPUS__=8 -O2 -I../../bpf/ -I./cmock  nat-test.c cmock/unity.c cmock/cmock.c ../../bpf/mock/mock_helpers.c ../../bpf/mock/mock_helpers_skb.c ../../bpf/mock/mock_conntrack_stub.c -o nat-test
clang: error: no such file or directory: '../../bpf/mock/mock_conntrack_stub.c'
make: *** [Makefile:35: nat-test] Error 1
make: Leaving directory '/home/kkourt/src/cilium-trees/cilium-cmock/test/bpf'

I'm probably missing something obvious but can't figure out what it is. What would be the proper way of running the tests?

brb commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,112 @@
+  struct ipv4_ct_tuple tuple;
+  struct ipv4_nat_entry state;
+  struct ipv4_nat_target target;
+
```
Would be good if you could add a small comment for each assertion explaining what it's testing under what conditions.

xinyuanzzz replied:
Added!

brb replied:
Cosmetics: we tend to follow the kernel coding standard in `bpf/`, so maybe you could break the lines into max 80 chars?

xinyuanzzz replied:
fixed!

brb commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,131 @@
+    struct ipv4_nat_target target;
+
+    mock_ct_lookup4_ExpectAnyArgsAndReturn(-1);
+    // ct_lookup4 returns a negative value so snat_v4_track_local will return
```
It would be good if you could explain under what circumstances the CT lookup returns -1 in real scenarios. Same goes to other comments.

xinyuanzzz replied:
>Could you split this PR into two commits - one for introducing the mocking infrastructure and friends, and other for introducing NAT tests?

Done!
> It would be good if you could explain under what circumstances the CT lookup returns -1 in real scenarios. Same goes to other comments.

Done!

brb reviewed (changes requested):
Could you split this PR into two commits - one for introducing the mocking infrastructure and friends, and other for introducing NAT tests?

xinyuanzzz committed:
ebpf unit testing -- 1/2: mocking infrastructure

hand-written files:
bpf/mock/bpf.yaml: specify CMock configuration options
bpf/mock/Makefile: contains rules to prepare for unit testing
bpf/mock/Dockerfile: contains commands to assemble the docker image
bpf/mock/helpers.sed: sed script file to generate header for helpers.h
bpf/mock/helpers_skb.sed: sed script file to generate header for helpers_skb.h
bpf/mock/helpers_xdp.sed: sed script file to generate header for helpers_xdp.h
bpf/mock/mock_helpers.sh: script to generate mock libraries for helper
functions
bpf/mock/mock_customized.sh: script to generate mock libraries for
customized functions
bpf/mock/README.md

modified files:
Makefile: add check_helper_headers in bpf/mock/Makefile to target lint

Signed-off-by: Xinyuan Zhang <zhangxinyuan@google.com>

joestringer commented on bpf/mock/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM ubuntu
```
```suggestion
FROM docker.io/library/ubuntu:20.04
```

joestringer commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,77 @@
+	sed -f helpers_skb.sed ../include/bpf/helpers_skb.h > helpers_skb.h; \
+	sed -f helpers_xdp.sed ../include/bpf/helpers_xdp.h > helpers_xdp.h;
+
+mock_helpers:
```
I hit this error when I pull these changes:

```
$ make -C bpf/mock mock_helpers                                                                                                  
open cmock.tar: no such file or directory
Unable to find image 'cmock:latest' locally
docker: Error response from daemon: pull access denied for cmock, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
Error: No such container:path: helpers:/mocks/
Error: No such container: helpers
Unable to find image 'cmock:latest' locally
docker: Error response from daemon: pull access denied for cmock, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
Error: No such container:path: helpers_skb:/mocks/
Error: No such container: helpers_skb
Unable to find image 'cmock:latest' locally
docker: Error response from daemon: pull access denied for cmock, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
Error: No such container:path: helpers_xdp:/mocks/
Error: No such container: helpers_xdp
mv: cannot stat 'mocks/*': No such file or directory
rmdir: failed to remove 'mocks': No such file or directory
make: *** [Makefile:49: mock_helpers] Error 1
```

I think that we can resolve this by following the Documentation/Makefile examples a bit more closely. For instance:
* Import the relevant top-level makefiles so that the makefile variables like `CONTAINER_ENGINE` can be reused
* The "builder-image" target in Documentation/ prepares the image in the way that "build_docker_image" target in this PR tries to do. This makefile can do very similar commands, maybe just tagging this as cilium/bpf-mock instead of the cilium/docs-builder.
* The `DOCKER_CTR` environment variable then runs this container, making the local directiory available within the container without having to manually `docker cp` each file in. The `env` from the docker version of this is not necessary here (READTHEDOCS / SKIP_LINT are only relevant to Documentation/). But you could basically copy the rest of this variable declaration into this makefile + DOCKER_RUN so this can be reused in other targets.
* Targets like `mock_helpers` should depend on the target to build the image, so that we know it is available before trying to run commands.
* Rather than individually running each mock command as a dedicated line in the Makefile, we can refactor this into a dedicated script which is made available inside the container, so that there is only one docker run command, and it invokes the script; the script can be super simple like `Documentation/check-cmdref.sh`, basically just a couple of lines which in this case would be the `cmock ruby ...` commands.

xinyuanzzz replied:
Thanks! Revised exactly as you suggest.

joestringer reviewed (changes requested):
Great progress :+1: . I've added some feedback around the docker file and make targets.

brb commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,94 @@
+
+    // If ct_lookup4 finds an entry, it will return a positive value. We can
+    // also assume it to be 1 because the actually value does not matter.
+    mock_ct_lookup4_ExpectAnyArgsAndReturn(1);a
```
The `a` at the end is a leftover.

xinyuanzzz replied:
Fixed!

brb commented on test/bpf/nat-test.c:
```diff
@@ -0,0 +1,23 @@
+int main(int argc, char *argv[])
+{
+    test_snat_v4_track_local();
+    return 0;
```
There was another test case previously. Any reason for the removal?

xinyuanzzz replied:
There was a test on function `snat_v4_new_mapping` which uses only mock helper functions. It was written only as a demo and not complete. So I removed it and keep only `test_snat_v4_track_local` which is a comprehensive test.

joestringer commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,74 @@
+	# Generate mock libraries for helpers.h, helpers_skb.h and helpers_xdp.h with CMock.
+	$(QUIET)$(DOCKER_RUN) ./mock_helpers.sh; \
+	mv mocks/* ./; \
+        rmdir mocks;
```
The indentation seems different here, probably because it's using spaces instead of tabs. Makefiles should use tabs, I'm not sure whether it guarantees to run the commands if the indentation is made differently.

joestringer commented on bpf/mock/Makefile:
```diff
@@ -0,0 +1,74 @@
+	# Generate the corresponding mock library.
+	$(QUIET)$(DOCKER_RUN) ./mock_customized.sh $(filename); \
+	mv mocks/* ./; \
+	rmdir mocks;
```
I think it's mostly a style nit, but it's a bit odd to delimit each of these lines with `; \`. This is not usually necessary for commands in a Make target. Maybe you picked this up because of the conditional statements in the `check_helper_headers` target, which are not proper commands for a makefile but you can get away with including them if you delimit those lines with `; \` because it somehow subtly swaps to running the entire line as a shell command rather than attempting to run the program at the start of the line and pass the rest of the arguments.

I don't think it makes a big difference so I'm not too fussed about changing this :shrug:, I just figured you might want to know.

xinyuanzzz replied:
Thanks for reminding me. I think it will be a little redundant to add more sh files to be called in the Makefile. Maybe we can leave them like this?

joestringer commented on bpf/mock/README.md:
````diff
@@ -0,0 +1,87 @@
+the Makefile, run
+
+```bash
+make -C test/bpf nat-test
````
I tried running this from the top-level directory, and it wouldn't work:

```
cilium $ make -C test/bpf nat-test
  CC    test/bpf/nat-test
clang: error: no such file or directory: '../../bpf/mock/mock_conntrack_stub.c'
make: *** [Makefile:35: nat-test] Error 1
```

xinyuanzzz replied:
That's because you need to run `make -C bpf/mock mock_customized filename=conntrack_stub.h` first to generate mock library for the customized functions used in nat.h. Maybe I should put this into `nat-test` in the Makefile?

Weil0ng replied:
I think it's at least worth adding a note here so people know they need to generate the mock lib first for this test to run.

joestringer replied:
Is there something stopping this from being a regular make dependency so the developer doesn't have to do this?

I appreciate the desire to provide a way that developers writing new tests can individually build / run specific tests. At the same time, I'd like to see this testing overall get to a state where automated systems can generate, build and run all of the tests automatically so that they don't regress. We can't rely on developers remembering to periodically run these tests, they will just bitrot and then the next developer will avoid them because they're broken.

joestringer replied:
Looks like the above should be achieved as long as we resolve this comment:

https://github.com/cilium/cilium/pull/16862#pullrequestreview-720427380

joestringer commented on test/bpf/Makefile:
```diff
@@ -30,6 +30,10 @@ elf-demo.o: elf-demo.c
 
+nat-test: nat-test.c $(LIB)
+	@$(ECHO_CC)
+	$(CLANG) ${FLAGS} -I../../bpf/ -I./cmock  $< cmock/unity.c cmock/cmock.c ../../bpf/mock/mock_helpers.c ../../bpf/mock/mock_helpers_skb.c ../../bpf/mock/mock_conntrack_stub.c -o $@
```
This builds the `nat-test` binary, but nothing is yet running the test.

There's already a `unit-test` target implied by lines 22 and 29, but could you maybe add one more small change to add this test to the regular suite?

One way to do this would be something like the following incremental patch:

```
$ git di
diff --git a/Makefile b/Makefile
index bffa7194572b..450233b82784 100644
--- a/Makefile
+++ b/Makefile
@@ -199,7 +199,6 @@ init-coverage: ## Initialize converage report for Cilium unit-tests.
 unit-tests: start-kvstores ## Runs all unit-tests for Cilium.
        $(QUIET) $(MAKE) $(SUBMAKEOPTS) -C tools/maptool/
        $(QUIET) $(MAKE) $(SUBMAKEOPTS) -C test/bpf/
-       test/bpf/unit-test
 ifeq ($(SKIP_VET),"false")
        $(MAKE) govet
 endif
diff --git a/test/bpf/Makefile b/test/bpf/Makefile
index f49afbc081aa..fadc02c85b1f 100644
--- a/test/bpf/Makefile
+++ b/test/bpf/Makefile
@@ -19,8 +19,10 @@ CLANG ?= $(QUIET) clang
 LLC ?= llc
 
 BPF_TARGETS := elf-demo.o
-TARGETS := $(BPF_TARGETS) unit-test
-all: $(TARGETS)
+ALL_TESTS := unit-test #nat-test
+TARGETS := $(BPF_TARGETS) $(ALL_TESTS)
+
+all: $(TARGETS) unit-tests
 
 elf-demo.o: elf-demo.c
        @$(ECHO_CC)
@@ -34,6 +36,12 @@ nat-test: nat-test.c $(LIB)
        @$(ECHO_CC)
        $(CLANG) ${FLAGS} -I../../bpf/ -I./cmock  $< cmock/unity.c cmock/cmock.c ../../bpf/mock/mock_helpers.c ../../bpf/mock/mock_helpers_skb.c ../../bpf/mock/mock_conntrack_stub.c -o $@
 
+unit-tests: $(ALL_TESTS)
+       @$(ECHO_CHECK)
+       $(foreach test,$<,$(ROOT_DIR)/$(RELATIVE_DIR)/$(test))
+
 clean:
        @$(ECHO_CLEAN)
        -$(QUIET)rm -f $(TARGETS)
+
+.PHONY: all unit-tests clean
```

I commented out unit-test in `$ALL_TESTS` above since it won't compile for me. Probably just need to replace the `../..` relative paths in the above command with instead `$(ROOT_DIR)`.

With this change, you can try it out also and confirm that this will be run from the CI by running `make -C test/bpf` and confirming that all the bpf unit tests build and run.

xinyuanzzz replied:
Revised as you suggested!

joestringer reviewed (changes requested):
I couldn't get the new test to compile or run just yet, so I made a few small suggestions in the right direction. Could you try those out and make sure that it's possible to run the new unit tests from the top-level of the git repository?

kkourt commented on Makefile:
```diff
@@ -476,9 +475,10 @@ govet: ## Run govet on Go source files in the repository.
+lint: ## Run golangci-lint and check if the helper headers in bpf/mock are up-to-date.
 	@$(ECHO_CHECK) golangci-lint
 	$(QUIET) golangci-lint run
+	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C bpf/mock/ check_helper_headers
```
hm...

Seems to me that if someone typed `make lint`, they wouldn't expect to run something related to `bpf/mock`. Do we maybe need another target for this?

joestringer replied:
I agree it's not the most obvious place, but at least this will get automatically run by the existing github actions.

Do you have a concrete proposal for somewhere else and how that should be hooked into the github actions?

(I don't disagree, but also I feel like we can always improve this iteratively if it's a problem for developers)

joestringer replied:
^^ There was also some background earlier in the PR, I can't exactly find it but the idea is that when someone updates the bpf helpers, we need to update also the generated mock helpers files, so it's sort-of like a linter in that regard.

xinyuanzzz replied:
`the idea is that when someone updates the bpf helpers, we need to update also the generated mock helpers files, so it's sort-of like a linter in that regard`

You are right. The idea is that `check_helper_headers` will report an error if the helper headers are out-of-date, and developers should run `generate_helper_headers` to update them.

kkourt replied:
> (I don't disagree, but also I feel like we can always improve this iteratively if it's a problem for developers)

OK, sounds good!

kkourt commented on Makefile:
```diff
@@ -199,7 +199,6 @@ init-coverage: ## Initialize converage report for Cilium unit-tests.
 unit-tests: start-kvstores ## Runs all unit-tests for Cilium.
 	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C tools/maptool/
 	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C test/bpf/
-	test/bpf/unit-test
```
Can you please describe the motivation for removing this line?
`test/bpf/unit-test.c` still seems to exist and it also seems to have some tests which are not part of the cmock tests. Or am I missing something?

joestringer replied:
I had suggested this along with changes to `test/bpf/Makefile` that should cause this target to be run automatically.

xinyuanzzz replied:
The process to run all the tests is now inside target `unit-tests` in `test/bpf/Makefile`.

kkourt replied:
Ah apologies, I missed the earlier discussion and the `test/bpf/Makefile` part that @joestringer kindly pointed below.

kkourt reviewed (commented):
Thanks for creating the container! It makes it much easier to run this.
Indeed, I've ran it locally and other than a few warnings about redefining NULL, seems to work OK.

I have a few more comments/questions, so please have a look below.

joestringer commented on test/bpf/Makefile:
```diff
@@ -19,8 +19,10 @@ CLANG ?= $(QUIET) clang
+ALL_TESTS := unit-test nat-test
+TARGETS := $(BPF_TARGETS) $(ALL_TESTS)
+
+all: $(TARGETS) unit-tests
```
@kkourt this line + the `unit-tests` dependency on `ALL_TESTS` below -> depending on `unit-test` should allow us to remove the direct `unit-test` call from the top-level makefile (though please do review to make sure I'm not mistaken :) ).

joestringer commented on test/bpf/Makefile:
```diff
@@ -30,6 +32,15 @@ elf-demo.o: elf-demo.c
+
+unit-tests: $(ALL_TESTS)
+	@$(ECHO_CHECK)
+	$(foreach test,$<,$(ROOT_DIR)/$(RELATIVE_DIR)/$(test))
```
Hmm, I thought this should work, but when I try it now that nat-test is in the $ALL_TESTS, it doesn't seem like the nat test is getting run. The iteration doesn't work the way I expected (I expected `$<` to expand to `unit-test nat-test` and then this iteration would run each of the binaries).

```
$ remake -x -C test/bpf unit-tests
Reading makefiles...
remake: Entering directory '/home/joe/git/cilium-review/test/bpf'
Updating makefiles...
Updating goal targets...
 File 'unit-tests' does not exist.
Must remake target 'unit-tests'.
Makefile:41: update target 'unit-tests' due to: unit-test nat-test
##>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
echo "  CHECK test/bpf"
##<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
  CHECK test/bpf
##>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
/home/joe/git/cilium-review/test/bpf/unit-test
##<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Successfully remade target file 'unit-tests'.
remake: Leaving directory '/home/joe/git/cilium-review/test/bpf'
```

joestringer replied:
^^ OK, I suggested the wrong variable. `$<` is the first dependency, we actually need `$^`:

https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html

joestringer replied:
I have reasonable confidence that this should work correctly:

```
$ git di
diff --git a/test/bpf/Makefile b/test/bpf/Makefile
index 24671d888f72..6526e3e28446 100644
--- a/test/bpf/Makefile
+++ b/test/bpf/Makefile
@@ -39,7 +39,10 @@ nat-test: nat-test.c $(LIB)
 
 unit-tests: $(ALL_TESTS)
        @$(ECHO_CHECK)
-       $(foreach test,$<,$(ROOT_DIR)/$(RELATIVE_DIR)/$(test))
+       for test in $^; do \
+               $(ECHO_CHECK) $$test; \
+               $(ROOT_DIR)/$(RELATIVE_DIR)/$$test; \
+       done
 
 clean:
        @$(ECHO_CLEAN)
```

xinyuanzzz replied:
Thanks! I just tried this and it worked. Updated!

kkourt reviewed (approved):
🚀

brb commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,94 @@
+    // matter.
+    mock_ct_lookup4_ExpectAnyArgsAndReturn(CT_NEW);
+    mock_ct_create4_ExpectAnyArgsAndReturn(-1);
+    // So snat_v4_track_local will return that value.
```
From the comments it's a bit not clear what this test case is actually testing. In particular, you could elaborate what it means if `snat_v4_track_local()` returns 0.

brb commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,94 @@
+    // return 0.
+    mock_ct_lookup4_ExpectAnyArgsAndReturn(CT_NEW);
+    mock_ct_create4_ExpectAnyArgsAndReturn(0);
+    // So snat_v4_track_local will return 0.
```
Same here.

joestringer commented on bpf/mock/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+RUN apt-get -y install ruby-full
+RUN apt-get -y install git
+RUN git clone --recursive https://github.com/ThrowTheSwitch/CMock.git
+RUN git -C CMock reset --hard 3d4ba8d20b8958da5dace7dd5d31155c94b60819
```
:thinking: do we still need two different dockerfiles to facilitate the overall process?

Even if we need to run docker multiple times, if the image is the same (as `test/bpf/Dockerfile`) then is it possible to add just one dockerfile and share the build / tag steps?

xinyuanzzz replied:
That makes a lot of sense. Addressed!

xinyuanzzz committed:
ebpf unit testing -- 2/2: mock helper libraries and NAT tests

hand-written files:
bpf/mock/conntrack_stub.h: header of functions to be mocked in conntrack.h
bpf/tests/nat_test.h: sample unit test program for some functions in nat.h
test/bpf/nat-test.c: contains main functions to run test functions nat_test.h

modified files:
test/bpf/Makefile: add target for the customized test

Signed-off-by: Xinyuan Zhang <zhangxinyuan@google.com>

joestringer reviewed (approved):
LGTM at a high level. I didn't try out the latest iteration but I did check the Travis output from the previous one and it looked like the test ran successfully there. So we should be able to discover any failures or regressions as they come up. I also didn't look at the details of the nat test.

Weil0ng commented on bpf/tests/nat_test.h:
```diff
@@ -0,0 +1,98 @@
+#include "lib/conntrack.h"
+
+// Define macros like the followings to make sure the original customized
+// functions are mapped to the mock functions.
```
This needs more explanation, for example, if I'm writing a new test, how do I know names of the mocked user function? Will it always be generated in the format of "mock_<func>"?

xinyuanzzz replied:
Thanks for pointing it out. Explanation added to both `bpf/tests/nat_test.h` and `bpf/mock/conntrack_stub.h`.

borkmann reviewed (approved):
Thanks, are you planning to add more follow-ups with further test coverage? Would be great to see.
